### ISSUE TITLE
False positive of "reportUnnecessaryComparison" with condition inside loop

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#6764]
**Describe the bug**
See sample code below. `f0` and `f1` are basically the same except for the swapped operands around `!=`. `f0` type-checks but `f1` gets reported.

Note:
Initial value of `a` needs to be `None` to have this issue appear.

**Code or Screenshots**

```python
def f0():
    a = None
    b = ''

    while True:
        if a != b:
            a = b


def f1():
    a = None
    b = ''

    while True:
        if b != a:  # False report.
            a = b
```

[Pyright Playground](https://pyright-play.net/?pyrightVersion=1.1.292&pythonVersion=3.12&strict=true&locale=en&code=CYUwZgBGAMAUCUAuAsAKAhiBDCBeCAcgPYB2IamEARnhAOR1oWYDuAFgJYA2IEAKgCcAriBTpKmDpBwBCfFTESl2WlSao0oSGACMCRZhz5iZZhhr4G6yu269BIg0qnUIc7IgwBiCADEsXADOvAIgAA5EAgAuAHRmyhhG1GhAA)

**VS Code extension or command-line**

I'm using Pylance with Pyright 1.1.338, and the same issue can be found with all versions supported by Pyright Playground (1.1.292~1.1.341).

**Comment by erictraut** [comment]
Thanks for reporting. I'm not able to repro it the playground. I was able to repro it locally — but only intermittently. It appears that the type evaluation depends on the order in which expressions are evaluated. Types should be consistent independent of evaluation order, so this is definitely the sign of a bug.

I'm able to repro the problem consistently by modifying the file (e.g. by typing a space at the bottom) while jiggling my mouse over the `a = b` statement. This causes the `a = b` statement to be evaluated first (in response to a hover text language server request).

I'll investigate further.

**Comment by Azureblade3808** [comment]
If `a` has an initial value of `None`, this issue can be consistently repro-ed.

**Comment by erictraut** [comment]
Yes, I'm able to repro consistently with an initial value of `None`. Thanks, that will make it easier to diagnose.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.343, which I just published. It will also be included in a future release of pylance.
